README
$Id: README,v 1.10 2006/03/17 19:55:32 obecker Exp $

This module contains Java source files for Joost.

The file build.xml is the Ant buildfile. Apache Ant is a Java-based build tool
and can be downloaded from http://ant.apache.org/
Required version: 1.6 or above

If you don't have Ant and just want to use your preferred Java IDE then
check out the section at the end of this README.

For compiling you need several jars in the lib subdirectory or in your
classpath respectively. The CVS contains only joostGen.jar, compiled from 
the sources in the helper module.

Additionally you have to install
- Commons Discovery from http://jakarta.apache.org/commons/discovery/
- Commons Logging from http://jakarta.apache.org/commons/logging/
  (Tested with 1.0.3 and 1.0.4)
- log4j from http://jakarta.apache.org/log4j/
  (Latest tested version: 1.2.9)
- junit.jar from http://junit.org/ (version 3.8.1) and
  xercesImpl.jar (version 2.6.0 or above) from http://xml.apache.org/xerces2-j/
  for compiling the src/test package
- fop.jar and avalon-framework-cvs-20020806.jar
  from http://xml.apache.org/fop/ 
  (Latest tested version: 0.20.5)
  for compiling sf.net.joost.emitter.FOPEmitter.java
  (avalon-framework-api-4.1.5.jar and avalon-framework-impl-4.1.5.jar work as well)
  If you don't want a FOP connection just remove this file and
  comment out some lines in sf.net.joost.Main.java
- Jakarta BSF from http://jakarta.apache.org/bsf/
- Some scripting engine which is to be used by the joost:script element.
  Check the BSF site for references which languages are supported and where to 
  get the libraries from.

In case you don't have at least Java 1.4 installed on your computer, you have to 
add manually a SAX2 compliant parser and the JAXP 1.1 interfaces, too.


This software is released under the Mozilla Public License Version 1.1,
see http://www.mozilla.org/MPL/



===========================================================================


Creating Java source files in the net.sf.joost.grammar package without Ant
--------------------------------------------------------------------------

Normally Ant calls some tools to create the missing source files. 
This section describes how to do these steps manually.
I assume that you are in a Unix environment. On a Windows platform 
just change the slashes to backslashes.

cd src/net/sf/joost/grammar
java -cp ../../../../../lib/joostGen.jar JLex.Main Yylex
java -cp ../../../../../lib/joostGen.jar net.sf.joost.helper.Preprocess Expr.cup ExprParser.cup
java -cp ../../../../../lib/joostGen.jar net.sf.joost.helper.Preprocess Pattern.cup PatternParser.cup
  in ExprParser.cup: replace manually "@@@START-NT@@@" with "Expr"
  in PatternParser.cup: replace manually "@@@START-NT@@@" with "Pattern"
java -cp ../../../../../lib/joostGen.jar java_cup.Main -parser PatternParser -symbols Sym -runtime net.sf.joost.grammar.cup -interface -compact_red -nowarn PatternParser.cup
java -cp ../../../../../lib/joostGen.jar java_cup.Main -parser ExprParser -symbols Sym -runtime net.sf.joost.grammar.cup -interface -compact_red -nowarn ExprParser.cup
cd ../../../../..

You only need to repeat these steps if you change one of the files Yylex,
Expr.cup, Pattern.cup, terminals.inc or expr.inc.


CHANGES
Changes since 0.9.1
- Bugfix: 2930934
- XMLReader features can be specified via system props on the command line using
  -DsaxFeature-<feature URL>=[true|false]

Changes since 20090315
- Bugfixes: 2840109, 2840121, 2840583, 2843559
- changed the net.sf.joost.TransformerHandlerResolver interface such that both
  resolve(..) methods will now receive both a javax.xml.transform.ErrorListener
  parameter and a javax.xml.transform.URIResolver parameter (fixes 2843559)

Changes since 20081012
- Bugfixes: 2403950, 2687662, 2687675

Changes since 20080528
- Bugfixes: 1993357, 1993925, 1994290, 2144242, 2157146
- added disable-output-escaping support (mainly for XSLT filter processes)

Changes since 20080309
- Bugfixes: 1928572, 1951834, 1951836, 1975067

Changes since 20070718
- Bugfixes: 1758424, 1810643, 1824901, 1824909, 1899453
- added stx:doctype
- replaced attribute 'case' of stx:analyze-text with 'flags'
   * case="insensitive" must be specified now using flags="i"
   * multiline mode is now by default off and must be enabled via flags="m"
- API: the default constructor of net.sf.joost.trax.TransformerFactoryImpl does
  not longer throw an IOException
- API: removed class net.sf.joost.plugins.traxfilter.XalanReusableTH 
  and field net.sf.joost.plugins.traxfilter.THResolver.USE_INTERNAL_XALAN_TH

Changes since 20070621
- added interface OutputURIResolver (Feature request 1170466)
- the binary should be Java 1.4 compatible again
- Bugfix: 1755658

Changes since 20070529
- Bugfixes: 1730697, 1741087
- incorporated code from Michael Kay's Saxon to support full XPath 2.0
  regular expressions
- added matches(), replace(), tokenize()

Changes since 20070527
- added stx:analyze-text, stx:match, stx:no-match, sf:regex-group

Changes since 20061005
- Bugfixes: 1574978, 1689706
- Optimization when processing the attributes of literal result elements

Changes since 20060505
- Bugfixes: 1489567, 1509455

Changes since 20060205
- Added special element joost:script for script support (e.g. Javascript)
  (thanks again to Fikin).
- Bugfix: 1446221

Changes since 20050521
- added plugin ability for net.sf.joost.TransformerHandlerResolver classes;
  no need any more to programatically register filter implementations, Joost
  will automatically detect them after their registration as Java 1.3 services,
  i.e. in the META-INF/services/net.sf.joost.TransformerHandlerResolver file
  (thanks to Fikin).
- xslt and stx filters support transformers reusage (thanks to Fikin).
- added support for Unicode surrogate pairs
- output numbers in a XML Schema conformant way (prevent scientific notation)
  (Feature request 1337383)
- added 'append' attribute to stx:result-document
- Fix for #1215468 required a change in the TransformerHandlerResolver interface:
  added URIResolver parameter in 
  TransformerHandler resolve(String, String, String, URIResolver, Hashtable)
- Bugfixes: 1215468, 1365639

Changes since 20041107
- added command line switch -noext and TrAX attribute ALLOW_EXTENSION_FUNCTIONS
  to prevent/allow the call of Java extension functions
- changed url(...) specification for the 'filter-src' attribute:
  the part between 'url(' and ')' will be evaluated as an STXPath expression.
  This means unquoted URLs like in "url(filename.xsl)" won't work anymore,
  use "url('filename.xsl')" instead.
- relaxed buffer restrictions: allow adding non-well-formed XML within a
  single stx:result-buffer instruction; the buffer contents must be
  well-formed when stx:process-buffer is called (thanks to Maurice Jumelet)
- Bugfixes: 1086639, 1086657, 1093383, 1107878, 1170453, 1170454, 1170668

Changes since 20041002
- Joost now requires Java 1.4 (or above)
- generate character references for characters that can't be represented
  in the current encoding
- renamed attribute 'stylesheet-prefix' to 'sheet-prefix' in 
  stx:namespace-alias
- renamed value '#stylesheet' to '#sheet' for the 'base' attribute of
  stx:process-document
- changed Joost's extension namespace to "http://joost.sf.net/extension"
- added experimental "joost:html" output method (in that extension namespace)
- added command line switch -nodecl
- added 'terminate', 'logger', and 'level' attributes to stx:message
- made Commons Logging usage optional
- Bugfix: 1056611
  
Changes since 20040330
- boosted Joost's performance about 25% 
  (major internal refactoring contributed by Thomas Behrends)
- added 'select' attribute to stx:message, stx:comment, and 
  stx:processing-instruction
- added filter-method "http://www.ietf.org/rfc/rfc2616.txt#POST"
- flush output on error
- Bugfixes: 936354, 1003790

Changes since 20031219
- added command line switches: -time, -m
- Bugfixes: 869207, 890961, 925215

Changes since 20030914
- allow nested XPath comments (: ... :)
  (yet not covered by the STX spec)
- added escape-uri()
- added 'output-method' attribute to stx:transform and stx:result-document,
  added support for text output (spec change 2003-12-08)
- renamed attribute 'encoding' to 'output-encoding' in stx:result-document
  (dito)
- Changes according to WD-xpath-functions-20031112:
  * renamed get-in-scope-namespaces() to get-in-scope-prefixes()
  * swapped arguments of get-namespace-uri-for-prefix()
  * don't drop NaNs in sum(), avg(), min(), max()

src/net/sf/joost/Main.java
/*
 * $Id: Main.java,v 1.32 2010/01/12 22:29:28 obecker Exp $
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is: this file
 *
 * The Initial Developer of the Original Code is Oliver Becker.
 *
 * Portions created by  ______________________
 * are Copyright (C) ______ _______________________.
 * All Rights Reserved.
 *
 * Contributor(s): Nikolay Fiykov
 */

package net.sf.joost;

import net.sf.joost.emitter.FOPEmitter;
import net.sf.joost.emitter.StreamEmitter;
import net.sf.joost.emitter.StxEmitter;
import net.sf.joost.stx.ParseContext;
import net.sf.joost.stx.Processor;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

import javax.xml.transform.OutputKeys;
import javax.xml.transform.SourceLocator;
import javax.xml.transform.TransformerException;

import org.apache.commons.logging.Log;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.log4j.PropertyConfigurator;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

/**
 * Command line interface for Joost.
 * @version $Revision: 1.32 $ $Date: 2010/01/12 22:29:28 $
 * @author Oliver Becker
 */
public class Main implements Constants
{
   // the logger object if available
   private static Log log = OptionalLog.getLog(Main.class);

   /**
    * Entry point
    * @param args array of strings containing the parameter for Joost and
    * at least two URLs addressing xml-source and stx-sheet
    */
   public static void main(String[] args)
   {
      // input filename
      String xmlFile = null;

      // the currently last processor (as XMLFilter)
      Processor processor = null;

      // output filename (optional)
      String outFile = null;

      // custom message emitter class name (optional)
      String meClassname = null;

      // log4j properties filename (optional)
      String log4jProperties = null;

      // log4j message level (this is an object of the class Level)
      Level log4jLevel = null;

      // set to true if a command line parameter was wrong
      boolean wrongParameter = false;

      // set to true if -help was specified on the command line
      boolean printHelp = false;

      // set to true if -pdf was specified on the command line
      boolean doFOP = false;

      // set to true if -nodecl was specified on the command line
      boolean nodecl = false;

      // set to true if -noext was specified on the command line
      boolean noext = false;

      // set to true if -doe was specified on the command line
      boolean doe = false;

      // debugging
      boolean dontexit = false;

      // timings
      boolean measureTime = false;
      long timeStart = 0, timeEnd = 0;

      // needed for evaluating parameter assignments
      int index;

      // serializer SAX -> XML text
      StreamEmitter emitter = null;

      // filenames for the usage and version info
      final String USAGE = "usage.txt", VERSION = "version.txt";

      try {

         // parse command line argument list
         for (int i=0; i<args.length; i++) {
            if (args[i].trim().length() == 0) {
               // empty parameter? ingore
            }
            // all options start with a '-', but a single '-' means stdin
            else if (args[i].charAt(0) == '-' && args[i].length() > 1) {
               if ("-help".equals(args[i])) {
                  printHelp = true;
                  continue;
               }
               else if ("-version".equals(args[i])) {
                  printResource(VERSION);
                  logInfoAndExit();
               }
               else if ("-pdf".equals(args[i])) {
                  doFOP = true;
                  continue;
               }
               else if ("-nodecl".equals(args[i])) {
                  nodecl = true;
                  continue;
               }
               else if ("-noext".equals(args[i])) {
                  noext = true;
                  continue;
               }
               else if ("-doe".equals(args[i])) {
                  doe = true;
                  continue;
               }
               else if ("-wait".equals(args[i])) {
                  dontexit = true; // undocumented
                  continue;
               }
               else if ("-time".equals(args[i])) {
                  measureTime = true;
                  continue;
               }
               else if ("-o".equals(args[i])) {
                  // this option needs a parameter
                  if (++i < args.length && args[i].charAt(0) != '-') {
                     if (outFile != null) {
                        System.err.println("Option -o already specified with "
                                           + outFile);
                        wrongParameter = true;
                     }
                     else
                        outFile = args[i];
                     continue;
                  }
                  else {
                     if (outFile != null)
                        System.err.println("Option -o already specified with "
                                           + outFile);
                     else
                        System.err.println("Option -o requires a filename");
                     i--;
                     wrongParameter = true;
                  }
               }
               else if ("-m".equals(args[i])) {
                  // this option needs a parameter
                  if (++i < args.length && args[i].charAt(0) != '-') {
                     if (meClassname != null) {
                        System.err.println("Option -m already specified with "
                                           + meClassname);
                        wrongParameter = true;
                     }
                     else
                        meClassname = args[i];
                     continue;
                  }
                  else {
                     if (meClassname != null)
                        System.err.println("Option -m already specified with "
                                           + meClassname);
                     else
                        System.err.println("Option -m requires a classname");
                     i--;
                     wrongParameter = true;
                  }
               }
               else if(DEBUG && "-log-properties".equals(args[i])) {
                  // this option needs a parameter
                  if (++i < args.length && args[i].charAt(0) != '-') {
                     log4jProperties = args[i];
                     continue;
                  }
                  else {
                     System.err.println("Option -log-properties requires " +
                                        "a filename");
                     wrongParameter = true;
                  }
               }
               else if (DEBUG && "-log-level".equals(args[i])) {
                  // this option needs a parameter
                  if (++i < args.length && args[i].charAt(0) != '-') {
                     if ("off".equals(args[i])) {
                        log4jLevel = Level.OFF;
                        continue;
                     }
                     else if ("debug".equals(args[i])) {
                        log4jLevel = Level.DEBUG;
                        continue;
                     }
                     else if ("info".equals(args[i])) {
                        log4jLevel = Level.INFO;
                        continue;
                     }
                     else if ("warn".equals(args[i])) {
                        log4jLevel = Level.WARN;
                        continue;
                     }
                     else if ("error".equals(args[i])) {
                        log4jLevel = Level.ERROR;
                        continue;
                     }
                     else if ("fatal".equals(args[i])) {
                        log4jLevel = Level.FATAL;
                        continue;
                     }
                     else if ("all".equals(args[i])) {
                        log4jLevel = Level.ALL;
                        continue;
                     }
                     else {
                        System.err.println("Unknown parameter for -log-level: "
                                           + args[i]);
                        wrongParameter = true;
                        continue;
                     }
                  }
                  else {
                     System.err.println("Option -log-level requires a " +
                                        "parameter");
                     wrongParameter = true;
                  }
               }
               else {
                  System.err.println("Unknown option " + args[i]);
                  wrongParameter = true;
               }
            }
            // command line argument is not an option with a leading '-'
            else if ((index = args[i].indexOf('=')) != -1) {
               // parameter assignment
               if (processor != null)
                  processor.setParameter(args[i].substring(0,index),
                                        args[i].substring(index+1));
               else {
                  System.err.println("Assignment " + args[i] +
                                     " must follow an stx-sheet parameter");
                  wrongParameter = true;
               }
               continue;
            }
            else if (xmlFile == null) {
               xmlFile = args[i];
               continue;
            }
            else {
               // xmlFile != null, i.e. this is an STX sheet
               ParseContext pContext = new ParseContext();
               pContext.allowExternalFunctions = !noext;
               if (measureTime)
                  timeStart = System.currentTimeMillis();
               Processor proc =
                  new Processor(new InputSource(args[i]), pContext);
               if (nodecl)
                  proc.outputProperties.setProperty(
                        OutputKeys.OMIT_XML_DECLARATION, "yes");
               if (measureTime) {
                  timeEnd = System.currentTimeMillis();
                  System.err.println("Parsing " + args[i] + ": " +
                                     (timeEnd - timeStart) + " ms");
               }

               if (processor != null)
                  proc.setParent(processor); // XMLFilter chain
               processor = proc;
            }
         }

         // PDF creation requested
         if (doFOP && outFile == null) {
            System.err.println("Option -pdf requires option -o");
            wrongParameter = true;
         }

         // missing filenames
         if (!printHelp && processor == null) {
            if (xmlFile == null)
               System.err.println("Missing filenames for XML source and " +
                                  "STX transformation sheet");
            else
               System.err.println("Missing filename for STX transformation " +
                                  "sheet");
            wrongParameter = true;
         }

         if (meClassname != null && !wrongParameter) {
            // create object
            StxEmitter messageEmitter = null;
            try {
               messageEmitter =
                  (StxEmitter)Class.forName(meClassname).newInstance();
            }
            catch (ClassNotFoundException ex) {
               System.err.println("Class not found: " +
                                  ex.getMessage());
               wrongParameter = true;
            }
            catch (InstantiationException ex) {
               System.err.println("Instantiation failed: " +
                                  ex.getMessage());
               wrongParameter = true;
            }
            catch (IllegalAccessException ex) {
               System.err.println("Illegal access: " +
                                  ex.getMessage());
               wrongParameter = true;
            }
            catch (ClassCastException ex) {
               System.err.println("Wrong message emitter: " +
                                  meClassname + " doesn't implement the " +
                                  StxEmitter.class);
               wrongParameter = true;
            }
            if (messageEmitter != null) { // i.e. no exception occurred
               // set message emitter for all processors in the filter chain
               Processor p = processor;
               do {
                  p.setMessageEmitter(messageEmitter);
                  Object o = p.getParent();
                  if (o instanceof Processor)
                     p = (Processor)o;
                  else
                     p = null;
               } while (p != null);
            }
         }

         if (printHelp) {
            printResource(VERSION);
            printResource(USAGE);
            logInfoAndExit();
         }

         if (wrongParameter) {
            System.err.println("Specify -help to get a detailed help message");
            System.exit(1);
         }

         if (DEBUG) {
            // use specified log4j properties file
            if (log4jProperties != null)
               PropertyConfigurator.configure(log4jProperties);

            // set log level specified on the the command line
            if (log4jLevel != null)
               Logger.getRootLogger().setLevel(log4jLevel);
         }


         // The first processor re-uses its XMLReader for parsing the input
         // xmlFile.
         // For a real XMLFilter usage you have to call
         // processor.setParent(yourXMLReader)

         // Connect a SAX consumer
         if (doFOP) {
            // pass output events to FOP
//              // Version 1: use a FOPEmitter object as XMLFilter
//              processor.setContentHandler(
//                 new FOPEmitter(
//                    new java.io.FileOutputStream(outFile)));

            // Version 2: use a static method to retrieve FOP's content
            // handler and use it directly
            processor.setContentHandler(
               FOPEmitter.getFOPContentHandler(
                  new java.io.FileOutputStream(outFile)));
         }
         else {
            // Create XML output
            if (outFile != null) {
               emitter = StreamEmitter.newEmitter(outFile,
                                                  processor.outputProperties);
               emitter.setSystemId(new File(outFile).toURI().toString());
            }
            else
               emitter = StreamEmitter.newEmitter(System.out,
                                                  processor.outputProperties);
            processor.setContentHandler(emitter);
            processor.setLexicalHandler(emitter);
            // the previous line is a short-cut for
            // processor.setProperty(
            //    "http://xml.org/sax/properties/lexical-handler", emitter);

            emitter.setSupportDisableOutputEscaping(doe);
         }

         InputSource is;
         if (xmlFile.equals("-")) {
            is = new InputSource(System.in);
            is.setSystemId("<stdin>");
            is.setPublicId("");
         }
         else
            is = new InputSource(xmlFile);

         // Ready for take-off
         if (measureTime)
            timeStart = System.currentTimeMillis();

         processor.parse(is);

         if (measureTime) {
            timeEnd = System.currentTimeMillis();
            System.err.println("Processing " + xmlFile + ": " +
                               (timeEnd - timeStart) + " ms");
         }

//           // check if the Processor copy constructor works
//           Processor pr = new Processor(processor);
//           java.util.Properties props = new java.util.Properties();
//           props.put("encoding", "ISO-8859-2");
//           StreamEmitter em =
//              StreamEmitter.newEmitter(System.err, props);
//           pr.setContentHandler(em);
//           pr.setLexicalHandler(em);
//           pr.parse(is);
//           // end check

         // this is for debugging with the Java Memory Profiler
         if (dontexit) {
            System.err.println("Press Enter to exit");
            System.in.read();
         }

      }
      catch (java.io.IOException ex) {
         System.err.println(ex.toString());
         System.exit(1);
      }
      catch (SAXException ex) {
         if (emitter != null) {
            try {
               // flushes the internal BufferedWriter, i.e. outputs
               // the intermediate result
               emitter.endDocument();
            }
            catch (SAXException exx) {
               // ignore
            }
         }
         Exception embedded = ex.getException();
         if (embedded != null) {
            if (embedded instanceof TransformerException) {
               TransformerException te = (TransformerException)embedded;
               SourceLocator sl = te.getLocator();
               String systemId;
               // ensure that systemId is not null; is this a bug?
               if (sl != null && (systemId = sl.getSystemId()) != null) {
                  // remove the "file://" scheme prefix if it is present
                  if (systemId.startsWith("file://"))
                     systemId = systemId.substring(7);
                  else if (systemId.startsWith("file:"))
                     // bug in JDK 1.4 / Crimson? (see rfc1738)
                     systemId = systemId.substring(5);
                  System.err.println(systemId + ":" +
                                     sl.getLineNumber() + ":" +
                                     sl.getColumnNumber() + ": " +
                                     te.getMessage());
               }
               else
                  System.err.println(te.getMessage());
            }
            else {
               // Fatal: this mustn't happen
               embedded.printStackTrace(System.err);
            }
         }
         else
            System.err.println(ex.toString());
         System.exit(1);
      }
   }


   /**
    * Outputs the contents of a resource info file.
    * @param filename the name of the file containing the info to output
    */
   private static void printResource(String filename)
   {
      try {
         // find the file resource
         InputStream is = Main.class.getResourceAsStream(filename);
         if (is == null)
            throw new java.io.FileNotFoundException(filename);
         BufferedReader br = new BufferedReader(new InputStreamReader(is));
         boolean doOutput = true;
         String line;
         while ((line = br.readLine()) != null) {
            if (line.startsWith("@@@ ")) { // special control line
               if (line.equals("@@@ START DEBUG ONLY"))
                  doOutput = DEBUG;
               else if (line.equals("@@@ END DEBUG ONLY"))
                  doOutput = true;
               // else: ignore
               continue;
            }
            if (doOutput)
               System.err.println(line);
         }
         System.err.println("");
      }
      catch (IOException ex) {
         if (log != null)
            log.error(ex);
         else
            System.err.println(ex);
      }
   }


   /**
    * Output logging availability info and exit Joost
    */
   private static void logInfoAndExit() {
      System.err.println("Logging is "
                         + ((log != null)
                               ? "enabled using " + log.getClass().getName()
                               : "disabled"));
      System.exit(0);
   }
}
